controllers/owner_and_name.go

```diff
@@ -0,0 +1,96 @@
+package controllers
+
+import (
+	"context"
+	"errors"
+
+	"github.com/go-logr/logr"
+	commonAnnotations "github.com/medik8s/common/pkg/annotations"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/openshift/api/machine/v1beta1"
+
+	"github.com/medik8s/self-node-remediation/api/v1alpha1"
+)
+
+// GetNodeName gets the node name:
+// - if owned by NHC, or as fallback, from annotation or CR name
+// - if owned by a Machine, from the Machine's node reference
+func GetNodeName(ctx context.Context, c client.Client, snr *v1alpha1.SelfNodeRemediation, log logr.Logger) (string, error) {
+	// NHC has priority, so check it first: in case the SNR is owned by NHC, get the node name from annotation or CR name
+	if ownedByNHC, _ := IsOwnedByNHC(snr); ownedByNHC {
+		return getNodeNameDirect(snr), nil
+	}
+	// in case the SNR is owned by a Machine, we need to check the Machine's nodeRef
+	if ownedByMachine, ref := IsOwnedByMachine(snr); ownedByMachine {
+		return getNodeNameFromMachine(ctx, c, ref, snr.GetNamespace(), log)
+	}
+	// fallback: annotation or name
+	return getNodeNameDirect(snr), nil
+}
+
+func getNodeNameDirect(snr *v1alpha1.SelfNodeRemediation) string {
+	nodeName, isNodeNameAnnotationExist := snr.GetAnnotations()[commonAnnotations.NodeNameAnnotation]
+	if isNodeNameAnnotationExist {
+		return nodeName
+	}
+	return snr.GetName()
+}
+
+// IsOwnedByNHC checks if the SNR CR is owned by a NodeHealthCheck CR.
+func IsOwnedByNHC(snr *v1alpha1.SelfNodeRemediation) (bool, *metav1.OwnerReference) {
+	for _, ownerRef := range snr.OwnerReferences {
+		if ownerRef.Kind == "NodeHealthCheck" {
+			return true, &ownerRef
+		}
+	}
+	return false, nil
+}
+
+// IsOwnedByMachine checks if the SNR CR is owned by a Machine CR.
+func IsOwnedByMachine(snr *v1alpha1.SelfNodeRemediation) (bool, *metav1.OwnerReference) {
```

Comment:
Nit: this could be made private

Reply:
leftover from an earlier version, when it also was in a different package